Add support for case search label multimedia

mkangia opened the pull request:
## Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
Best to review by commit.

[Ticket](https://dimagi-dev.atlassian.net/browse/USH-901)
[Requirements](https://dimagi-dev.atlassian.net/browse/USH-901?focusedCommentId=139013)

The second and final PR after https://github.com/dimagi/commcare-hq/pull/29735 to add support for multimedia for case search labels.
(This PR can be reviewed without reading the first PR)

Summary for reviewer for main commits:
This PR just builds on same lines on how we add multimedia for Case List,
First commit,
1. inherit from `NavMenuItemMediaMixin`, add a way to get_app and add reference for `_module` in `assign_references`
2. add media references in view context, with identifier, same as that passed to the template
3. add case search labels medias to module's `all_media` just like case list media. I believe this would make them available in downloads and other areas where multimedia files are accessed for a module.

Second commit,
1. fetch paths set for different media in view and save it 

Following commits are self explanatory.

For reference,
Code changes exploration done earlier in https://github.com/dimagi/commcare-hq/pull/29714

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
`USH_CASE_CLAIM_UPDATES`

## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
Users can now set image and audio multimedia for case search labels "Search" and "Search Again".

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
https://dimagi-dev.atlassian.net/browse/QA-3055

### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
Tested locally. Will also go through QA.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

mkangia committed:
setup UI for search labels multimedias

mkangia committed:
save set multimedias

mkangia committed:
fix doc

mkangia committed:
include labels multimedias in rename media

mkangia commented on corehq/apps/app_manager/templates/app_manager/partials/modules/case_search_properties.html:
```diff
@@ -119,6 +119,11 @@ <h4 class="panel-title panel-title-nolink">{% trans "Search and Claim Options" %
             </div>
           </div>
+          <div class="case-search-multimedia-input">
+            <!-- ko stopBinding: true -->
```
stopBinding used because the enclosing template with bound to a different object

https://github.com/dimagi/commcare-hq/blob/8469ce862df7561c3cf86205f6a9f49804fdf45e/corehq/apps/app_manager/templates/app_manager/partials/modules/case_search_properties.html#L9

and these are to be bound through a different mechanism

https://github.com/dimagi/commcare-hq/blob/5ca2ffe38023b0412c906d1aac1830bf0e4fa032/corehq/apps/app_manager/static/app_manager/js/nav_menu_media.js#L4-L13

orangejenny replied:
As with most self-PR comments, this would be a good code comment (maybe without the code references).

orangejenny replied:
Alternatively, what would you think of adding a param to `nav_menu_media.html` to optionally include the `stopBindings`? Our custom knockout components do something similar; they have a flag indicating whether they should apply their own bindings or not.

mkangia replied:
I think its better this way so that one can see in the parent template itself that a section of it is not binding along with the parent. If we move it within the nested template, that visibility would go.
Though, I will be happy to make the change to pass a param if it seems better.

orangejenny replied:
I think it can go either way, happy to defer to your judgment.

mkangia commented on corehq/apps/app_manager/views/view_generic.py:
```diff
@@ -193,6 +193,22 @@ def _make_name(suffix):
                 'qualifier': 'case_list-menu_item_',
             })
+            if (module and hasattr(module, 'search_config') and module.uses_media()
+                    and toggles.USH_CASE_CLAIM_UPDATES.enabled(domain)):
```
taking hints from how `get_case_list_menu_item_media` used a few lines above, is defined

https://github.com/dimagi/commcare-hq/blob/c46b51ca256aa4bc0d4cdeefe21f453bb199b17f/corehq/apps/hqmedia/models.py#L831-L837

the method for menu_refs, `get_case_search_label_media` would be
```
def get_case_search_label_media(self, module, label, to_language=None):
    if not module or not module.uses_media() or not hasattr(module, 'search_config'):
        return {}
    # rest of the code
```
So i just thought of getting those conditions up here itself.
If these specific media are not added there really should not be any issue since the UI would not be showing the media fields as well.

mkangia commented on corehq/apps/app_manager/views/modules.py:
```diff
@@ -1114,8 +1114,14 @@ def edit_module_detail_screens(request, domain, app_id, module_unique_id):
         ):
             search_label = module.search_config.search_label
             search_label.label[lang] = search_properties.get('search_label', '')
+            search_label._set_media("media_image", lang, search_properties.get('search_label_image'))
```
this method is called at another place already, considering making it public (removing the `_`).

orangejenny replied:
I agree with removing the `_`.

mkangia replied:
resolved by https://github.com/dimagi/commcare-hq/pull/29765/commits/89838899561e2de0f4a306ae4392c5ed8f1944e2

mkangia commented on corehq/apps/hqmedia/models.py:
```diff
@@ -672,6 +672,11 @@ def rename_media(self, old_path, new_path):
+        # Case search and claim menu items
+        if hasattr(self, 'search_config'):
+            count += self.rename_menu_media(self.search_config.search_label, old_path, new_path)
+            count += self.rename_menu_media(self.search_config.search_again_label, old_path, new_path)
```
accidentally stumbled on this method while searching for `hasattr(self, 'case_list')` :)

orangejenny replied:
Seems like it's part of ICDS's multimedia paths upload/download. I'm tempted to say we should remove that whole feature, but it looks like USH is using it.

mkangia commented on corehq/apps/app_manager/static/app_manager/js/details/screen_config.js:
```diff
@@ -1109,7 +1109,7 @@ hqDefine('app_manager/js/details/screen_config', function () {
                             containsCaseListLookupConfiguration: (columnType === "short" && (hqImport('hqwebapp/js/toggles').toggleEnabled('CASE_LIST_LOOKUP') || hqImport('hqwebapp/js/toggles').toggleEnabled('BIOMETRIC_INTEGRATION'))),
                             // TODO: Check case_search_enabled_for_domain(), not toggle. FB 225343
-                            containsSearchConfiguration: (columnType === "short" && hqImport('hqwebapp/js/toggles').toggleEnabled('SYNC_SEARCH_CASE_CLAIM')),
+                            containsSearchConfiguration: (columnType === "short" && hqImport('hqwebapp/js/toggles').toggleEnabled('USH_CASE_CLAIM_UPDATES')),
```
https://github.com/dimagi/commcare-hq/pull/29735#discussion_r634715474
makes me think this is the correct flag.

It was originally set in https://github.com/dimagi/commcare-hq/pull/11438/files

orangejenny replied:
I don't think this is right. This flag controls the saving of the entire case search model: https://github.com/dimagi/commcare-hq/blob/0c8d6d5f404beca1a92d29ea8d0f1ff4e3ee95cf/corehq/apps/app_manager/static/app_manager/js/details/screen_config.js#L996-L998

That model contains SYNC_SEARCH_CASE_CLAIM features, not just USH_CASE_CLAIM_UPDATES features.

mkangia replied:
Thanks. Will revert that.

is the understanding that any domain that would have `USH_CASE_CLAIM_UPDATES` would also have `SYNC_SEARCH_CASE_CLAIM` for sure?
Because if that is not then I am quite confused.

mkangia replied:
this was the latest commit so i simply removed it for cleaner history

orangejenny replied:
Yeah, any domain with `USH_CASE_CLAIM_UPDATES` turned on also has to have `SYNC_SEARCH_CASE_CLAIM` turned on.

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/details/case_claim.js:
```diff
@@ -203,6 +203,9 @@ hqDefine("app_manager/js/details/case_claim", function () {
         };
 
         subscribeToSave(self, searchConfigKeys, saveButton);
+        $(".case-search-multimedia-input button").on("click", function () {
```
If someone is using a custom multimedia path, and they just change the path, will the save button get triggered? There might be a generic handler that listens for changes on all `<input>` elements, I can't remember.

mkangia replied:
Good catch. Changing just the path does not enable save. I will take a look.

mkangia replied:
resolved by https://github.com/dimagi/commcare-hq/pull/29765/commits/1541eb65b5f80484d7e6ffd9336ef72589465119

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/details/case_claim.js:
```diff
@@ -215,7 +215,15 @@ hqDefine("app_manager/js/details/case_claim", function () {
                 search_additional_relevant: self.searchAdditionalRelevant(),
                 search_button_display_condition: self.searchButtonDisplayCondition(),
                 search_label: self.searchLabel(),
+                search_label_image:
```
Are the `|| null` parts necessary?

mkangia replied:
I took that from other similar implementations.
Seems like the reason is that otherwise this would be empty string in case nothing is set. So that is being modified to pass null instead. I am not sure if that makes a different later but going to assume it was done for a reason i have not found that.

orangejenny reviewed (commented):
The comment on `containsSearchConfiguration` is blocking, the rest is minor.

mkangia committed:
make method public

mkangia committed:
be more specific for labels

mkangia committed:
save default label mm fields

mkangia committed:
add media references to build files

orangejenny committed:
Moved attributes up to ActionMixin

orangejenny committed:
Added previously Action-only attributes to LocalizedAction

mkangia commented:
Hey @orangejenny 
thanks for bringing this up to date. The first commit you made here, is same as in the [PR](https://github.com/dimagi/commcare-hq/pull/29816) you made. Should it be only here considering its only in this PR we are adding LocalizedAction?

mkangia committed:
trigger save for checkboxes

orangejenny commented:
@mkangia Practically, it doesn't matter, since the bug won't be exposed until this PR is merged, but in theory it's a bug with the old code, so I'd like to get it fixed and merged.

Could you add tests to this PR?

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/details/case_claim.js:
```diff
@@ -203,9 +203,14 @@ hqDefine("app_manager/js/details/case_claim", function () {
             saveButton.fire('change');
         });
+        // checkbox to select media for all languages
+        $(".case-search-multimedia-input input[type='checkbox']").on('click', function () {
```
I think this should be `change` because `click` won't notice if a user toggles the checkbox with their space bar.

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/29765/commits/05a49f1d95e86b554b7679c6e29fa45428e3e240

I would not have thought of anyone using spacebar + tabs to get to the checkbox. took me a few tabs to get to the checkbox. :)

mkangia committed:
add documentation

mkangia committed:
use change instead of click

mkangia committed:
fix error when fetching default translations

mkangia committed:
add tests for case search app strings

mkangia committed:
better name

mkangia committed:
add tests for case search actions in suite files

mkangia commented:
Hey @orangejenny 

thanks for your last review.

> Could you add tests to this PR?

Added test for the suite files to account for Action/LocalizedAction change.

Also wrote test for app strings, that revealed a bug, which I fixed in https://github.com/dimagi/commcare-hq/pull/29765/commits/d65e6b4d1c66717c1afb3ef5b15da43ccae9e850

Requesting hopefully last round of review on this. This is ready to merged now since the production migration for labels is complete.

orangejenny commented on corehq/apps/app_manager/tests/test_suite.py:
```diff
@@ -150,6 +152,48 @@ def test_sort_cache_suite(self, *args):
+            "./detail[@id='m0_search_short']/action"
+        )
+
+        # test for localized action node for new versions
```
"new" versions 😉

mkangia replied:
should that be newer? :package:

mkangia replied:
updated to better documentation and also to trigger a test build :)

mkangia committed:
update doc

mkangia committed:
Merge branch 'master' into mk/case_search_labels/multimedia

mkangia committed:
assign references where case search added

mkangia committed:
reset module as well

mkangia committed:
add default sort nodes

mkangia commented on corehq/apps/app_manager/tests/data/suite/search_command_detail.xml:
```diff
@@ -16,6 +16,11 @@
           <xpath function="case_name"/>
         </text>
       </template>
+      <sort direction="ascending" order="1" type="string">
```
@orangejenny can you please help confirm if this is because of https://github.com/dimagi/commcare-hq/pull/29784/files?

I checked out to master, and added
```
# wrap to have assign_references called
self.app = Application.wrap(self.app.to_json())
# reset to newly wrapped module
self.module = self.app.modules[0]
```
in `setUp`, that gave me two sort blocks in the diff which makes me think that wrapping the application doc brings sort setting into action, though here there was a diff of 4, which is not clear why that would be.

mkangia replied:
found the real reason on why this was happening here, fixing it here https://github.com/dimagi/commcare-hq/pull/29866.

Its expected to have these 4 sort blocks added after wrap
